File: UniBlock/src/main/scala/persistence/PersistenceManager.scala
package it.unifi.nave.uniblock
package persistence

import crypto.PKHelper
import data.EventContainer
import data.event.Event.EventType
import data.event.{Certificate, Event}
import persistence.impl.InMemoryPersistence

import java.io.{ByteArrayInputStream, ObjectInputStream}
import java.util.Base64

object PersistenceManager {
  private val persistenceManager = InMemoryPersistence

  def blockchain: Blockchain = persistenceManager

  def keyManager: KeyManager = persistenceManager

  def searchCertificate(id: String): Certificate = searchCertificate(id, Event.Certificate)

  private def searchCertificate(id: String, eventType: EventType): Certificate = blockchain
    .flatMap(_.eventContainers)
    .find(e => e.eventType == eventType && e.author == id)
    .map(deserializeCertificate)
    .get

  private def deserializeCertificate(eventContainer: EventContainer): Certificate = {
    if (!verifyCertificate(eventContainer)) throw new IllegalArgumentException("Can't verify certificate")
    val ois = new ObjectInputStream(new ByteArrayInputStream(Base64.getDecoder.decode(eventContainer.payload)))
    ois.readObject.asInstanceOf[Certificate]
  }

  private def verifyCertificate(eventContainer: EventContainer): Boolean = eventContainer match {
    case EventContainer(_, Event.Genesis, _, _) => true
    case EventContainer(_, _, payload, sign) => PKHelper.verify(Right(payload), sign, searchGenesisCertificate.signPbk)
  }

  def searchGenesisCertificate: Certificate = searchCertificate("", Event.Genesis)
}

File: src/main/java/it/unifi/nave/uniblock/persistence/PersistenceManager.java
package it.unifi.nave.uniblock.persistence;

import com.google.common.collect.Streams;
import com.google.common.primitives.Bytes;
import it.unifi.nave.uniblock.data.block.Block;
import it.unifi.nave.uniblock.data.event.Certificate;
import it.unifi.nave.uniblock.helper.crypto.PKHelper;
import it.unifi.nave.uniblock.persistence.impl.InMemoryPersistence;

import java.security.PublicKey;
import java.util.Collection;

public class PersistenceManager {
    private static final InMemoryPersistence IN_MEMORY_PERSISTENCE = new InMemoryPersistence();

    public static Blockchain getBlockchain() {
        return IN_MEMORY_PERSISTENCE;
    }

    public static KeyManager getKeyManager() {
        return IN_MEMORY_PERSISTENCE;
    }

    public static Certificate searchCertificate(String userId) {
        return Streams.stream(getBlockchain())
                .map(Block::getEvents)
                .flatMap(Collection::stream)
                .filter(Certificate.class::isInstance)
                .map(Certificate.class::cast)
                .filter(c -> userId.equals(c.userId()))
                .filter(PersistenceManager::verifyCertificate)
                .findAny()
                .orElseThrow();
    }

    private static boolean verifyCertificate(Certificate certificate) {
        if (certificate.certificateType() == Certificate.CertificateType.GENESIS) {
            return true;
        } else {
            return verify(certificate.signPbk(), certificate.dhPbk(), certificate.sign());
        }
    }

    private static boolean verify(PublicKey signPbk, PublicKey dhPbk, String sign) {
        return PKHelper.verify(Bytes.concat(signPbk.getEncoded(), dhPbk.getEncoded()), sign, searchGenesisCertificate().signPbk());
    }

    public static Certificate searchGenesisCertificate() {
        return searchCertificate(Certificate.GENESIS);
    }

}